src/cockpit/packages.py

```diff
@@ -262,6 +262,27 @@ def check(self, at_least_prio):
             if self.manifest['priority'] <= at_least_prio:
                 return False
 
+        # this can be a single value, a list, or not exist at all
+        def get_condition_value(key):
+            value = self.manifest.get('conditions', {}).get(key)
+            if value is None:
+                return []
+            if isinstance(value, list):
+                return value
+            # single value
+            return [value]
+
+        CONDITIONS = {
+            'path-exists': os.path.exists,
+            'path-not-exists': lambda p: not os.path.exists(p),
+        }
+
+        for condition, test_fn in CONDITIONS.items():
+            for value in get_condition_value(condition):
+                if not test_fn(value):
+                    logger.debug('Hiding package %s as its %s %s condition is not met', self.path, condition, value)
+                    return False
```

Comment:
This is awfully complicated, and I'm not sure I understand what the purpose of it all is.  Do you have a case in mind where all of the conditions of a given type would need to be met?  What about one or more?

I could imagine, for example, a condition would be met if either `/usr/bin/x` *or* `/usr/local/bin/x` are available.

If you'd like to support multiple items to check for then I suggest you drop the "single item" mode (it's complicated and will be annoying to express the type of in the future) and replace the key name with something more descriptive like `"all-paths-exist"` and `"no-paths-exist"`.  It wasn't clear from me from reading the documentation, for example, if a list to `"path-exists"` meant "all" or "any", and even less clear for the `not-exists` case.

In any case, I'd pull the content of the (vastly simplified) nested function into the loop body in the next iteration of this patch.  Or just write it like:

```python3
if not all(os.path.exists(path) for path in conditions.get('all-paths-exist', [])):
    return False
if any(os.path.exists(path) for path in conditions.get('no-paths-exist', [])):
    return False
return True
```

Reply:
What I had in mind was something like "/dev/kvm and /usr/lib/systemd/libvirtd.service both exist" -- but at that point, it's all pretty much made up. As I said I took inspiration from systemd units, whose conditions are conjunctive.

However, agreed that ORed conditions can also make sense. But maybe both my `/usr/lib/systemd/ vs. /lib/systemd vs. /usr/local/lib/systemd` and your `/usr/bin/x or /usr/local/bin/x` are telling us that we should use different checks: `unit-exists` and `program-exists` (using `shutil.which()`)?

Adding "or" *could* be done with having lists instead of strings as entries, but then we'd again need to support both types:

```json
"paths-exist": [
    "/dev/kvm",
    ["/usr/local/lib/systemd/system/libvirtd.service", "/lib/systemd/system/libvirtd.service"]
]
```

but honestly this looks rather ridiculous and error prone. What I'd *really* like to do is to have a list of conditions, each of which have a predicate, like `[ ("path-exists", "/a"), ("unit-exists", "foo.service")]`, but that's not JSON -- requiring two-element lists could work.

So in summary, I don't yet have a brilliant idea about how to map propositional logic onto JSON..

Reply:
How about this: `"conditions": [ "path:/good", "!path:/bad", "unit:foo.service"]` ? It's not JSON down to the primitives, but not hard to parse and a lot easier to read? That still doesn't address the "or" support, but at least the readability and "single vs. multiple" structure.

Reply:
Once we go there, we could of course go one step further and `.split('|')` to support alternatives:

    "conditions": [ "path:/good | /also/good", "!path:/bad", "unit:packagekit.service | packagekitd.service"]

@marusak , @allisonkarlitskaya : Opinions?

(OMG, I'm gonna need  a lot of unit tests)

Reply:
I think this is getting more complicated, not less.

I'm also not fond of the pseudo-language embedded in JSON.  Just use JSON for that:

```json
"conditions": [
  "any": [{"path-exists": "foo"}, {"path-exists": "bar"}],
  "all": [ "..." ],
]
```

But really, this is getting extreme.  I'd start with:

```json
"conditions" [
  {"path-exists": "foo"},
  {"path-does-not-exist": "bar"},
]
```

and we can go Turing-complete if and when we find the actual need to do so.

Reply:
Implemented as above.